### ISSUE TITLE
Making levels more flexible in napari workflows

### ISSUE THREAD

**Issue by jluethi** [fractal-analytics-platform/fractal-tasks-core#159]
This is a follow-up to the initial implementation of supporting different pyramid levels in napari workflows as discussed in https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/139

For context, most workflows should probably run at level 0 and it's great we have optional flexibility to run e.g. segmentation at lower resolution.

### The current implementation
(correct me if I'm wrong with something here @tcompa):
1) A workflow is run at a global pyramid level (0, 1, 2, 3 etc.), defaulting to level 0
2) The output is saved at the same pyramid level as the input
3) All label images are upscaled to the resolution of the intensity images (handling the case in which a label image is lower resolution)
4) We have a check that enforces that only workflows that go from images to label images can be at level > 0.

**tldr: I think this implementation is good for what we need at the moment. Below are a few thoughts of things we may consider in the future**

### Areas we may want to adapt:
a) Currently, label images can only be loaded if there are corresponding intensity images to do the upscaling, right (see (3) above)? Which is a limitation we may not want to enforce. See here for the check: https://github.com/fractal-analytics-platform/fractal-tasks-core/blob/8dc8e192c561d8f605c07bbe6e0700294aa55ddc/fractal_tasks_core/napari_workflows_wrapper.py#L183
Why would we not enforce this? A user could do measurements just on the label images. A classical case would be a regionprops measurement of label images to get areas and other morphological measurements. We currently can measure them as part of a workflow that measures intensity statistics as well, but a user could define a regionprops workflow that only measures those morphology features and thus not require intensity images.  
b) We may want to change the hard check for which workflows can be run at level > 0 into a warning. I think it's a good default assumption, but there may be edge cases. But relaxing this could bring its own issues, see c)
c) Loading at level > 0 when labels and intensity images don't have the same resolution: This could get a bit tricky. But if I understand the current code correctly, we load with a level integer. E.g. let's say label is 4x downsampled from intensity image. A full intensity image is 2560x2160. A full label image is 640x540. If we request level 1, currently we get the intensity image at 1280x1080 and the label image at 320x270. It would make more sense though to load the label image at 640x540, because we anyway need to upscale to 1280x1080. Thus, loading with a level integer, not a resolution target, can be tricky.
=> @tcompa are there other places this may hit us? I don't think so at the moment, but may be good to be aware of the limitation of loading with level integers across labels & intensity images.

### Urgency of adaptations
a) is not urgent. Worst-case, an intensity image needs to be loaded and nothing is done with it. It's more of a niche case.
b) Not urgent. Most workflows should anyway run at level 0. Once someone brings a compelling use-case, let's think about relaxing this.
c) As long as we keep 4) / b) as an exception, not a warning, this doesn't come up

**Comment by jluethi** [comment]
@tcompa If there is an easy way to fix a) without having to mess with the label upscaling, let's do that. Otherwise, let's not worry about the adaptation ideas in here for the time being :)

**Comment by tcompa** [comment]
> ### The current implementation
> 
> (correct me if I'm wrong with something here @tcompa):
> 
>     1. A workflow is run at a global pyramid level (0, 1, 2, 3 etc.), defaulting to level 0
> 
>     2. The output is saved at the same resolution as the input
> 
>     3. All label images are upscaled to the resolution of the intensity images (handling the case in which a label image is lower resolution)
> 
>     4. We have a check that enforces that only workflows that go from images to label images can be at level > 0.

Agreed.

**Comment by tcompa** [comment]
> a) Currently, label images can only be loaded if there are corresponding intensity images to do the upscaling, right (see (3) above)? Which is a limitation we may not want to enforce. See here for the check:
> 
> https://github.com/fractal-analytics-platform/fractal-tasks-core/blob/8dc8e192c561d8f605c07bbe6e0700294aa55ddc/fractal_tasks_core/napari_workflows_wrapper.py#L183
> Why would we not enforce this? A user could do measurements just on the label images. A classical case would be a regionprops measurement of label images to get areas and other morphological measurements. We currently can measure them as part of a workflow that measures intensity statistics as well, but a user could define a regionprops workflow that only measures those morphology features and thus not require intensity images.

As per our brief call today, the only reason this requirement is in-place is that without an `image` input we cannot upscale a `label` input. New issue:
* #161

**Comment by tcompa** [comment]
> b) We may want to change the hard check for which workflows can be run at level > 0 into a warning. I think it's a good default assumption, but there may be edge cases. But relaxing this could bring its own issues, see c)

This takes a second to implement (from hard to soft constraint), I can do it whenever you prefer.

**Comment by jluethi** [comment]
Sounds good. Let's see if we can have label images without intensity images then. And let's keep this as the backlog for using napari workflows with levels > 0 for non-labeling workflows :)

> This takes a second to implement (from hard to soft constraint), I can do it whenever you prefer.

Good to know. I'm not sure yet whether this constraint should be relaxed because of the complexities in c). So let's wait to see if there are relevant use-cases for it first.

**Comment by tcompa** [comment]
> Good to know. I'm not sure yet whether this constraint should be relaxed because of the complexities in c). So let's wait to see if there are relevant use-cases for it first.

Agreed. And, of course, what takes a second is only replacing the constraint - not checking for consequences ;)

**Comment by tcompa** [comment]
@jluethi, can we close this one in favor of #161, or are is there anything left to be done here?

**Comment by jluethi** [comment]
Yes, let's reopen this if we want to tackle using levels > 0 for non-labeling workflows. The rest is in #161
